# GUÍA 1

**Nombre:** Marjoris Parejo Castillo  
**Fecha:** 17/08/2024  
**Programa:** Especialización en Machine Learning

## Actividad 1. Funciones y tipos de variables

In [245]:
#Importo las librerías que voy a utilizar para explorar la base de datos
from tabulate import tabulate
import pandas as pd

In [247]:
#Realizo la carga del archivo csv que se va a explorar 'Census Income'
df=pd.read_csv('census_income-1.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Unnamed: 0      48842 non-null  int64 
 1   age             48842 non-null  int64 
 2   workclass       47879 non-null  object
 3   fnlwgt          48842 non-null  int64 
 4   education       48842 non-null  object
 5   education-num   48842 non-null  int64 
 6   marital-status  48842 non-null  object
 7   occupation      47876 non-null  object
 8   relationship    48842 non-null  object
 9   race            48842 non-null  object
 10  sex             48842 non-null  object
 11  capital-gain    48842 non-null  int64 
 12  capital-loss    48842 non-null  int64 
 13  hours-per-week  48842 non-null  int64 
 14  native-country  48568 non-null  object
 15  income          48842 non-null  object
dtypes: int64(7), object(9)
memory usage: 6.0+ MB


In [249]:
#Visualizo el nombre de cada columna del DataFrame y el tipo de datos de cada una de ellas
df.dtypes

Unnamed: 0         int64
age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
income            object
dtype: object

In [251]:
#Quito la columna Unnamed: 0 del DataFrame para tener una mejor visualización ya que esta columna actua como indicador del número de fila
del df['Unnamed: 0']

**Descripción de las variables**  
**age:** Edad de la persona  
**workclass:** Tipo de empleo de la persona  
**fnlwgt:** Peso final, representando la cantidad de personas que el individuo representa en la población.  
**education:** Nivel educativo de la persona  
**education-num:** Número asociado al nivel educativo de la persona  
**marital-status:** Estado civil de la persona  
**occupation:** Ocupación de la persona  
**relationship:** Relación de la perosna individuo con el jefe de la familia.  
**race:** Raza de la persona  
**sex:** Género de la persona  
**capital-gain:** Ganancia de capital obtenida durante el año
**capital-loss:** Pérdida de capital sufrida durante el año
**hours-per-week:** Número de horas trabajadas por semana.  
**native-country:** País de origen de la persona  
**income:** Ingreso anual de la persona, indicando si gana más de 50,000 dólares (>50K) o 50,000 dólares o menos  

In [254]:
#Observo las primeras filas del DataFrame para verificar la carga inicial de datos y entender la estructura o formato de estos
df.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [256]:
#Observo las últimas filas del Dataframe para verificar la carga completa de datos y validar la estructura o formato
df.tail()

age     workclass  fnlwgt  education  education-num  \
48837   39       Private  215419  Bachelors             13   
48838   64           NaN  321403    HS-grad              9   
48839   38       Private  374983  Bachelors             13   
48840   44       Private   83891  Bachelors             13   
48841   35  Self-emp-inc  182148  Bachelors             13   

           marital-status       occupation    relationship  \
48837            Divorced   Prof-specialty   Not-in-family   
48838             Widowed              NaN  Other-relative   
48839  Married-civ-spouse   Prof-specialty         Husband   
48840            Divorced     Adm-clerical       Own-child   
48841  Married-civ-spouse  Exec-managerial         Husband   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
48837               White  Female             0             0              36   
48838               Black    Male             0             0              40   
48839               White    Male             0             0              50   
48840  Asian-Pac-Islander    Male          5455             0              40   
48841               White    Male             0             0              60   

      native-country  income  
48837  United-States  <=50K.  
48838  United-States  <=50K.  
48839  United-States  <=50K.  
48840  United-States  <=50K.  
48841  United-States   >50K.

In [258]:
#Observo una por una las variables y su tipo

#Voy a crear una lista para almacenar el tipo de datos de cada columna del DataFrame
d_columnas = []


#Utilizo la iteración para obtener el tipo de datos de cada columna del DataFrame
for columna in df.columns:
    #Obtengo el tipo de dato de la columna
    tipo_dato = df[columna].dtype
    #Agrego a la lista el nombre de la columna y el tipo de dato
    d_columnas.append([columna, tipo_dato])

#Visualizo los datos en el formato de tabla
print(tabulate(d_columnas, headers=['Columna', 'Tipo de dato'], tablefmt='grid'))


+----------------+----------------+
| Columna        | Tipo de dato   |
+================+================+
| age            | int64          |
+----------------+----------------+
| workclass      | object         |
+----------------+----------------+
| fnlwgt         | int64          |
+----------------+----------------+
| education      | object         |
+----------------+----------------+
| education-num  | int64          |
+----------------+----------------+
| marital-status | object         |
+----------------+----------------+
| occupation     | object         |
+----------------+----------------+
| relationship   | object         |
+----------------+----------------+
| race           | object         |
+----------------+----------------+
| sex            | object         |
+----------------+----------------+
| capital-gain   | int64          |
+----------------+----------------+
| capital-loss   | int64          |
+----------------+----------------+
| hours-per-week | int64    

In [260]:
#Quiero saber que cantidad de observaciones tengo para cada variable (No nulos)
df.count()

age               48842
workclass         47879
fnlwgt            48842
education         48842
education-num     48842
marital-status    48842
occupation        47876
relationship      48842
race              48842
sex               48842
capital-gain      48842
capital-loss      48842
hours-per-week    48842
native-country    48568
income            48842
dtype: int64

In [262]:
#Quiero saber si tengo valores nulos y cuantos tengo
df.isnull().sum()

age                 0
workclass         963
fnlwgt              0
education           0
education-num       0
marital-status      0
occupation        966
relationship        0
race                0
sex                 0
capital-gain        0
capital-loss        0
hours-per-week      0
native-country    274
income              0
dtype: int64

In [264]:
#Quiero conocer el número de valores distintos en cada columna
df.nunique()

age                  74
workclass             9
fnlwgt            28523
education            16
education-num        16
marital-status        7
occupation           15
relationship          6
race                  5
sex                   2
capital-gain        123
capital-loss         99
hours-per-week       96
native-country       42
income                4
dtype: int64

In [266]:
#Veo una lista con el nombre de las columnas
df.columns.tolist()

['age',
 'workclass',
 'fnlwgt',
 'education',
 'education-num',
 'marital-status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'capital-gain',
 'capital-loss',
 'hours-per-week',
 'native-country',
 'income']

Los tipos de cada variable son:  
**Numéricas:** 'age','fnlwgt','capital-gain', 'capital-loss y 'hours-per-week'  
**Categóricas** 'workclass''
 'education
 'education-nu,
 'marital-stat',
 'occupatn',
 'relationip',
 ace',
er-week',
 'nate-country  ',
 'income'

In [269]:
#Observo el resumen estadístico para observar el comportamiento de cada variable númerica
df.describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  48842.000000  4.884200e+04   48842.000000  48842.000000  48842.000000   
mean      38.643585  1.896641e+05      10.078089   1079.067626     87.502314   
std       13.710510  1.056040e+05       2.570973   7452.019058    403.004552   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.175505e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.781445e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.376420e+05      12.000000      0.000000      0.000000   
max       90.000000  1.490400e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    48842.000000  
mean        40.422382  
std         12.391444  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

A partir del resumen estadístico se pueden sacar las siguientes conclusiones:
- La edad de las personas se encuentra en el rango de 17 a 90, con una edad promedio de 38.64 años
- El 75% de las personas trabajan 45 horas o menos por semana y el restante trabajan más de 45
- La ganancia de capital promedio es 1,079.07, pero con una alta variabilidad (7,452.02), lo que sugiere que muchas personas tienen ganancias mínimas, mientras que algunas obtienen ganancias muy grandes
- La alta desviación estándar en 'capital-gain' indica una gran variación en las ganancias de capital entre las personas.

In [272]:
#Calculo la frecuencia de cada categoría de la variable income
df['income'].value_counts()

income
<=50K     24720
<=50K.    12435
>50K       7841
>50K.      3846
Name: count, dtype: int64

Como podemos observar, la variable *'income'* debería tomar los valores de *<=50K* ó *>50K* pero en algunas observaciones se incluye un punto al final,  
por lo tanto, el punto debe eliminarse para obtener la frecuencia total para cada categoría

In [275]:
#Elimino el punto final de las observaciones que lo contienen y obtengo la frecuencia de cada una de las dos categorías
df['income'].str.rstrip('.').value_counts()

income
<=50K    37155
>50K     11687
Name: count, dtype: int64

In [277]:
#Lo observo en valor porcentual para tener una mejor interpretación del comportamiento de la variable 'income'
df['income'].str.rstrip('.').value_counts(normalize=True)*100

income
<=50K    76.071823
>50K     23.928177
Name: proportion, dtype: float64

A partir de los valores porcentuales puedo decir que el **76.07%** de personas de la muestra estudiada ganan menos de $50.000 y el **23.93%** ganan más de esta cifra.

In [280]:
#Calculo la frecuencia de la variable 'sex'
df['sex'].value_counts()

sex
Male      32650
Female    16192
Name: count, dtype: int64

In [282]:
#Lo observo en valor porcentual para tener una mejor interpretación del comportamiento de la variable 'sex'
df['sex'].str.rstrip('.').value_counts(normalize=True)*100

sex
Male      66.848204
Female    33.151796
Name: proportion, dtype: float64

A partir de los valores porcentuales puedo decir que el **66.85%** de personas de la muestra estudiada son hombres y el **33.15%** son mujeres

## Actividad 2. Programa con IF para la limpieza de datos


In [286]:
#Realizo la carga del archivo csv que se va a explorar 'titanic'
df1=pd.read_csv('titanic-1.csv')
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 887 entries, 0 to 886
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Survived                 887 non-null    int64  
 1   Pclass                   887 non-null    int64  
 2   Name                     887 non-null    object 
 3   Sex                      887 non-null    object 
 4   Age                      887 non-null    float64
 5   Siblings/Spouses Aboard  887 non-null    int64  
 6   Parents/Children Aboard  887 non-null    int64  
 7   Fare                     887 non-null    float64
dtypes: float64(2), int64(4), object(2)
memory usage: 55.6+ KB


**Descripción de las variables**
- **Survived**: Indica si el pasajero sobrevivió
- **Pclass**: Clase del pasajero en el Titanic
- **Name**: Nombre completo del pasajero
- **Sex**: Género del pasajero
- **Age**: Edad del pasajero en años
- **Siblings/Spouses Aboard**:Número de hermanos y/o cónyuges a bordo del Titanic
- **Parents/Children Aboard**: Número de padres e hijos a bordo del Titanic
- **Fare**: Tarifa pagada por el pasajero

In [289]:
#Visualizo el nombre de cada columna del DataFrame y el tipo de datos de cada una de ellas
df1.dtypes

Survived                     int64
Pclass                       int64
Name                        object
Sex                         object
Age                        float64
Siblings/Spouses Aboard      int64
Parents/Children Aboard      int64
Fare                       float64
dtype: object

Las variables númericas según el tipo de dato son: 'Survived', 'Pclass', 'Age' y 'Fare'.
Las variables categóricas son: 'Name' y 'Sex'

In [292]:
#Visualizo las primeras observaciones del DataFrame para conocer su estructura
df1.head()

Survived  Pclass                                               Name  \
0         0       3                             Mr. Owen Harris Braund   
1         1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2         1       3                              Miss. Laina Heikkinen   
3         1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4         0       3                            Mr. William Henry Allen   

      Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  
0    male  22.0                        1                        0   7.2500  
1  female  38.0                        1                        0  71.2833  
2  female  26.0                        0                        0   7.9250  
3  female  35.0                        1                        0  53.1000  
4    male  35.0                        0                        0   8.0500

In [294]:
#Visualizo las últimas observaciones del DataFrame para validar la carga de datos completa y la estructura
df1.tail()

Survived  Pclass                            Name     Sex   Age  \
882         0       2            Rev. Juozas Montvila    male  27.0   
883         1       1     Miss. Margaret Edith Graham  female  19.0   
884         0       3  Miss. Catherine Helen Johnston  female   7.0   
885         1       1            Mr. Karl Howell Behr    male  26.0   
886         0       3              Mr. Patrick Dooley    male  32.0   

     Siblings/Spouses Aboard  Parents/Children Aboard   Fare  
882                        0                        0  13.00  
883                        0                        0  30.00  
884                        1                        2  23.45  
885                        0                        0  30.00  
886                        0                        0   7.75

In [296]:
#Cuento cuantas observaciones tiene cada variable (No nulas)
df1.count()

Survived                   887
Pclass                     887
Name                       887
Sex                        887
Age                        887
Siblings/Spouses Aboard    887
Parents/Children Aboard    887
Fare                       887
dtype: int64

In [298]:
total_rows = len(df1)

# Verifico si todas las columnas tienen el número máximo de observaciones (sin valores nulos)
if (df1.count() != total_rows).any():
    print("Hay columnas con valores nulos")
    print(df1.count())
else:
    print("Todas las columnas tienen el número máximo de observaciones (sin valores nulos).")

Todas las columnas tienen el número máximo de observaciones (sin valores nulos).


In [300]:
#Cuento cuantos valores nulos tiene cada variable
df1.isnull().sum()

Survived                   0
Pclass                     0
Name                       0
Sex                        0
Age                        0
Siblings/Spouses Aboard    0
Parents/Children Aboard    0
Fare                       0
dtype: int64

Ninguna de las variables tiene valores nulos

In [303]:
#Renombro columnas que contienen caracteres especiales para evitar errores de análisis posteriores
df1.rename(columns={'Siblings/Spouses Aboard': 'Siblings_Spouses_Aboard', 
                    'Parents/Children Aboard': 'Parents_Children_Aboard'}, inplace=True)

In [305]:
#Quiero saber cuales variables tienen 4 valores distintos o menos para ver su tipo y verificar que se están tomando de manera correcta
for column, count in df1.nunique().items():
    if count <= 4:
        # Obtengo el tipo de datos de la columna
        column_type = df1[column].dtype
        
        # Observo las columnas
        print(f"Columna: {column}")
        print(f"Número de valores distintos: {count}")
        print(f"Tipo de dato: {column_type}")
        print("-" * 40)

Columna: Survived
Número de valores distintos: 2
Tipo de dato: int64
----------------------------------------
Columna: Pclass
Número de valores distintos: 3
Tipo de dato: int64
----------------------------------------
Columna: Sex
Número de valores distintos: 2
Tipo de dato: object
----------------------------------------


De estas variables quiero analizar las dos numéricas: 'Survived' y 'Pclass ya que tienen 2 y 3 valores distintos respectivamente.

In [308]:
#Quiero saber cuales son los dos valores que toma la variable Survived y la frecuencia de estos
df1['Survived'].value_counts()

Survived
0    545
1    342
Name: count, dtype: int64

La variable 'Survived' dos valores: 0 y 1, lo que quiere decir que es una variable binaria.


In [311]:
#Quiero conocer que valores toma la variable Pclass y la frecuencia de estos
df1['Pclass'].value_counts()

Pclass
3    487
1    216
2    184
Name: count, dtype: int64

La variables 'Pclass' tomas tres valores: 1,2 y 3, que son las 3 clases de los pasajeros. Esto quiere decir que es una variable categórica

Teniendo en cuenta las observaciones de las variables 'Survived'y'Pclass' cambio su tipo a categoría

In [315]:
#Convierto las columnas 'Survived' y 'Pclass' de númericas a categóricas
df1['Survived'] = df1['Survived'].astype('category')
df1['Pclass'] = df1['Pclass'].astype('category')

#Para clarificar que 'Sex' es una variable categórica también le cambio el tipo
df1['Sex'] = df1['Sex'].astype('category')

#Verifico si el cambio fue exitoso
if df1['Survived'].dtype == 'category' and df1['Pclass'].dtype == 'category' and df1['Sex'].dtype == 'category':
    print("Columnas 'Survived' y 'Pclass' convertidas exitosamente de numéricas a categóricas.")
else:
    print("La conversión de las columnas 'Survived' y/o 'Pclass' no fue exitosa.")
df1.dtypes

Columnas 'Survived' y 'Pclass' convertidas exitosamente de numéricas a categóricas.


Survived                   category
Pclass                     category
Name                         object
Sex                        category
Age                         float64
Siblings_Spouses_Aboard       int64
Parents_Children_Aboard       int64
Fare                        float64
dtype: object

Después de los cambios realizados tenemos las variables clasificadas por tipo de la siguiente manera:  
**Númericas**: 'Age', 'Siblings/Spouses Aboard', 'Parents/Children Aboard' y 'Fare'  
**Categóricas**: 'Survived', 'Pclass' y 'Sex'  
**De texto**: 'Name'

In [318]:
#Realizo un resumen estadístico de las variables numéricas
#Verifico que solo se realice para variables numéricas
if not df1.select_dtypes(include=['float64', 'int64']).empty:
    print(df1.describe())

              Age  Siblings_Spouses_Aboard  Parents_Children_Aboard       Fare
count  887.000000               887.000000               887.000000  887.00000
mean    29.471443                 0.525366                 0.383315   32.30542
std     14.121908                 1.104669                 0.807466   49.78204
min      0.420000                 0.000000                 0.000000    0.00000
25%     20.250000                 0.000000                 0.000000    7.92500
50%     28.000000                 0.000000                 0.000000   14.45420
75%     38.000000                 1.000000                 0.000000   31.13750
max     80.000000                 8.000000                 6.000000  512.32920


A partir del resumen estadístico de las variables númericas 'Age','Siblings/Spouses Aboard','Parents/Children Aboard' y 'Fare'  
puedo anotar lo siguiente:
- La persona con menor edad en el Titanic tenía 0.42 años y la mayor 80 años, es decir, que habia desde bebés hasta adultos mayores en el Titánic.
- La edad promedio de los pasajeros era de 29.47 años
- La mayoría de los pasajeros viajaban solos o con un solo familiar, ya que el 50% no tenía hermanos ni esposos a bordo y el 75% tenía un máximo de uno.
- La tarifa promedio pagada por los pasajeros fue apróximadamente 32.31 libras esterlinas (£),pero la desviación estándar de 49.78 muestra una gran variación en los precios de los boletos, probablemente debido a las distintas clases de servicio disponibles.
- El 50% de los pasajeros tenía 28 años o menos, y el 75% tenía 38 años o menos, por tanto, se puede decir que había una gran cantidad de jóvenes a bordo.
- La tarifa mínima fue 0, lo que podría indicar errores o boletos gratuitos, mientras que la tarifa máxima de 512.33 sugiere que algunos pasajeros pagaron mucho, probablemente por viajar en primera clase.
- Algunos pasajeros viajaban con hasta 8 hermanos o esposos y hasta 6 padres o hijos, lo que sugiere la presencia de familias numerosas a bordo, aunque no eran muy comunes.